Emc/1033 carbonmark api to typescript

0xemc opened the pull request:
## Description

This PR involves migrating the carbonmark-api to typescript and the latest version of fastify & associated libraries. 

It involves a lift and shift of https://github.com/KlimaDAO/klimadao/tree/staging/carbonmark-api in to a new directory using the latest version of Fastify's CLI TypeScript template.

Unfortunately this PR has grown larger than expected and it has become difficult to isolate the typescript migration only (as it would require adding types that would later need to be removed by automated tools). I'm pushing this up for visibility and to begin the review process ASAP. @ladytrekker @cujowolf @Atmosfearful @Najada 

* Refactoring has been kept to a minimum where possible (though was unavoidable in some cases)
* Types are now generated from the graph endpoints using [graphql-codegen](https://the-guild.dev/graphql/codegen)
* Apollo Client has been removed in place of [graphql-request](https://www.npmjs.com/package/graphql-request). This allows us to generate a fully typed sdk to access all of the gql endpoints.
* `npm run test:postman` has been migrated to this project to ensure endpoints as a 200 response smoke test


### Todo
- [x] Implement changes made to https://github.com/KlimaDAO/carbonmark-api since migration. Base commit for this migration is ([da419a5a65d1e4ec08d518ca8d188c55c006f209](https://github.com/KlimaDAO/carbonmark-api/commit/da419a5a65d1e4ec08d518ca8d188c55c006f209))
- [x] Add a proper build task

## Related Ticket

Resolves #1033

ladytrekker commented:
@0xemc All builds are failing on vercel with this error:

<img width="851" alt="Bildschirmfoto 2023-05-16 um 12 16 18" src="https://github.com/KlimaDAO/klimadao/assets/95881624/8d9193eb-c353-467a-b6af-459eaf5c5538">

Did you run "npm run build-all" locally? Should be the same error, right?
Did you format the code too? "npm run format-all"

Atmosfearful reviewed (changes requested):
Looking good, it would be a good idea for me to triple check the auth/nonce stuff and think about how to write solid tests for it... (recommendations appreciated)

Also make sure to bring in my recent changes for `projects.js` and `users`.

Quick summary:
I format bignumbers in the API instead of the frontend, for `project`, `project.activities`, and for `user.activities` but not for other endpoints yet.
I also moved some fields into `project.stats`, renamed the fields in `project.prices` and deprecated `project.projectAddress`.

0xemc commented on carbonmark-api/README.md:
```diff
@@ -2,10 +2,11 @@
   </a>
 
 <h2 align="center">Carbonmark Api</h2>
+
```
ah yes great point, updated

0xemc commented on package.json:
```diff
@@ -27,7 +27,7 @@
     "dev-cms": "cd ./cms && npm run start",
     "dev-carbonmark": "concurrently --kill-others-on-fail \"npm:dev-lib\" \"npm run dev --workspace=carbonmark\"",
-    "dev-carbonmark-api": "npm run dev --workspace=carbonmark-api",
+    "dev-carbonmark-api": "npm run dev --workspace=carbonmark-api-ts",
```
It's not no, you're right. This is some left overs from early attempts I think.

0xemc commented on carbonmark-api/src/utils/marketplace.utils.ts:
```diff
@@ -0,0 +1,33 @@
+import { notNil } from "./functional.utils";
+
+export const updateListingUser =
+  (fb: FirebaseInstance) => async (listing: Partial<Listing>) => {
```
I hear you but I think the singleton approach might go against Fastify's encapsulation principal (https://fastify.dev/docs/latest/Reference/Encapsulation).

i.e. if we created a seperate Fastify instance it would share the same Firebase instance rather than a seperate instance. Not sure if this is an issue but seems cleaner to have a new one per instance? (especially with versioning coming up)

What do you think?

0xemc commented on carbonmark-api/src/utils/marketplace.utils.ts:
```diff
@@ -0,0 +1,33 @@
+    return { ...listing, seller };
+  };
+
+export const isListingActive = (listing: Partial<Listing>) =>
```
Yep fair enough, we can enforce this in the linter too via `@typescript-eslint/explicit-function-return-type`.

0xemc commented:
> Looking good, it would be a good idea for me to triple check the auth/nonce stuff and think about how to write solid tests for it... (recommendations appreciated)
> 
> Also make sure to bring in my recent changes for `projects.js` and `users`.
> 
> Quick summary: I format bignumbers in the API instead of the frontend, for `project`, `project.activities`, and for `user.activities` but not for other endpoints yet. I also moved some fields into `project.stats`, renamed the fields in `project.prices` and deprecated `project.projectAddress`.
> 
> When those changes are merged in here, we can drop the preview api URL for this PR into the carbonmark config in `carbonmark/lib/constants` and try running the frontend off it to make sure it all works. Then its good to go!

@Atmosfearful  By the way I've added an authentication flow test to this PR using `tap`. I'm thinking to migrate the postman tests in to the same format so it's consistent and we get coverage too.

Atmosfearful commented on carbonmark-api/src/.generated/types/marketplace.types.ts:
```diff
@@ -0,0 +1,1586 @@
+  Id = 'id'
+}
+
+export type Listing = {
```
This is great 💯 , we need a strategy for sharing the types between packages in the monorepo, and a process for ensuring they stay up to date.

For sharing, can we have a script that outputs these to `lib`? Or maybe we can safely import types (no runtime code) from `@klimadao/carbonmark-api` because all types get stripped out at compile time anyways...

For ensuring types stay up to date, perhaps the same approach we take with `prettier` where the build throws if the `format-all` exposes any uncommitted changes?
